helix-term/src/commands.rs

```diff
@@ -1645,9 +1645,123 @@ fn search_selection(cx: &mut Context) {
+
+                    let result_sink = sinks::UTF8(|line_num, _| {
+                        match all_matches_sx_cl
+                            .send((line_num as usize - 1, dent.path().to_path_buf()))
+                        {
+                            Ok(_) => Ok(true),
+                            Err(_) => Ok(false),
+                        }
+                    });
+                    let result = searcher_cl.search_path(&matcher_cl, dent.path(), result_sink);
+
+                    if let Err(err) = result {
+                        log::error!("Global search error: {}, {}", dent.path().display(), err);
+                    }
+                    WalkState::Continue
+                })
+            });
+
+        all_matches_rx.try_iter().collect()
+    })
+}
+
+/// returns a callback to display global search results
+fn global_search_callback(
+    all_matches: Vec<(usize, PathBuf)>,
+    current_path: Option<PathBuf>,
+) -> job::Callback {
+    Box::new(move |editor: &mut Editor, compositor: &mut Compositor| {
+        if all_matches.is_empty() {
+            editor.set_status("No matches found".to_string());
+            return;
+        }
+
+        let picker = FilePicker::new(
+            all_matches,
+            move |(_line_num, path)| {
+                let relative_path = helix_core::path::get_relative_path(path)
+                    .to_str()
+                    .unwrap()
+                    .to_owned();
+                if current_path.as_ref().map(|p| p == path).unwrap_or(false) {
+                    format!("{} (*)", relative_path).into()
+                } else {
+                    relative_path.into()
+                }
+            },
+            move |editor: &mut Editor, (line_num, path), action| {
+                match editor.open(path.into(), action) {
+                    Ok(_) => {}
+                    Err(e) => {
+                        editor.set_error(format!(
+                            "Failed to open file '{}': {}",
+                            path.display(),
+                            e
+                        ));
+                        return;
+                    }
+                }
+
+                let line_num = *line_num;
+                let (view, doc) = current!(editor);
+                let text = doc.text();
+                let start = text.line_to_char(line_num);
+                let end = text.line_to_char((line_num + 1).min(text.len_lines()));
+
+                doc.set_selection(view.id, Selection::single(start, end));
+                align_view(doc, view, Align::Center);
+            },
+            |_editor, (line_num, path)| Some((path.clone(), Some((*line_num, *line_num)))),
+        );
+        compositor.push(Box::new(picker));
+    })
+}
+
 fn global_search(cx: &mut Context) {
-    let (all_matches_sx, all_matches_rx) =
-        tokio::sync::mpsc::unbounded_channel::<(usize, PathBuf)>();
+    let (regex_sx, mut regex_rx) = tokio::sync::mpsc::channel(1);
```

Comment:
regex is sent from the callback of `Prompt` (` move |_view, _doc, regex, event| { ... regex_sx.send() ...}`), hence capturing `regex_sx`, I have just discovered that `oneshot::Sender` does not implement `Copy` trait and compiler would refuse to compile here.

Reply:
You can use a `move` closure to move it into the callback

Reply:
I have this error:
>error[E0507]: cannot move out of `regex_sx`, a captured variable in an `Fn` closure

I think the compiler cannot guarantee that we would call `regex_sx` only once, so it refuse to compile here.

Reply:
Right, using a channel seems like a hack to me anyway. I'd rather pass `cx` through the regex_prompt callback as a first parameter and manually call `let (doc, view) = current!(cx)`. Then you can create the search picker instance inside the callback.

Reply:
I think having a dedicated `FnOnce` callback for `PromptEvent::Validate` might clear things up a bit and we can use `oneshot` channel, but I really don't want to introduce any breaking change to the API, hence the hack.
 About passing `cx` through `regex_prompt`, wouldn't it mess up the lifetime of `cx` because we still need to `await` the search results (just my naive thought)?